### PR TITLE
Fix slice_strings wide strings logic with multi-byte characters

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes logic error in computing character and byte counts for slice positions in strings with specific pattern of multi-byte characters.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by jlowe — APPROVED** [review]
The Spark integration tests that were previously failing are now passing with this change.

File: cpp/tests/strings/slice_tests.cpp [hunk 1]
```diff
@@ -268,6 +268,23 @@ TEST_F(StringsSliceTest, MaxPositions)
...
+TEST_F(StringsSliceTest, MultiByteChars)
+{
+  auto input = cudf::test::strings_column_wrapper({
+    // clang-format off
+    "quick brown fox jumped over the lazy brown dog; the fat cats jump in place without moving "
+    "the following code snippet demonstrates how to use search for values in an ordered range  "
+    "it returns the last position where value could be inserted without the ééééé ordering ",
+    "algorithms execution is parallelized as determined by an execution policy; this is a 12345"
+    "continuation of previous row to make sure string boundaries are honored 012345678901234567"
+    "01234567890é34567890012345678901234567890"
```
Comment (vuule) [inline cpp/tests/strings/slice_tests.cpp:280]: why do we need this string to repro the issue? 
My guess: the issue is with `ééééé` and the rest of the string is there only to get to the size that triggers the execution policy that has the issue.

Comment (davidwendt) [reply cpp/tests/strings/slice_tests.cpp:280]: Yes. I can add some comments explaining this. 
The position of the `é` characters and the length of the strings verify logic fix in this PR.

Comment (vuule) [reply cpp/tests/strings/slice_tests.cpp:280]: A comment would be helpful, thank you!

**Review by zpuller — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Comment by sameerz** [comment]
Does this close https://github.com/rapidsai/cudf/issues/16768 ?

**Comment by davidwendt** [comment]
Yes, sorry I should have closed that with this merge.

**Pull request merged**